covert/elliptic/elligator.py

```diff
@@ -0,0 +1,206 @@
+# that point by q, it reduces to one of the eight *low order points* there are
+# in Curve25519. Standard public keys all reduce to big flat ZERO, where any
+# random point has equal chance of being in any of the eight sub groups, each
+# represented by its own low order point.
+#
+# The third problem needs to be solved by custom key generation that creates
+# *dirty points*, public keys that can be in any sub group, rather than only
+# in the prime group.
+#
+# We can make a point dirty by adding to it a random low point. This does not
+# affect the result of standard ECDH using that key, as any subgroups are
+# cancelled by the multiplication in those (because the secret scalars are
+# multiples of eight after the standard clamping required by the protocols)
+
+## High level API
+
+# Designed around Ed25519 because edsk can be easily converted into Montgomery
+# but the opposite is not possible. Also, starting with Ed25519 we have useful
+# extra bits to randomise the points, without having to create random numbers.
+# Curve25519 sk are already clamped, losing those crucial three low bits.
+
+# The sign of Ed25519 (lowest bit of x) is used as Elligator 2 "v sign", to
+# avoid having to calculate the v coordinate (which is rarely used anywhere),
+# but to still allow recovering the original Ed25519 public key exactly.
+
+from contextlib import suppress
+from secrets import token_bytes
+from typing import Tuple
+
+from .ed import LO, EdPoint, G, dirty_scalar
+from .scalar import fe, one, p, sqrtm1, zero
+from .util import sha, tobytes, toint
+
+
+class ElligatorError(ValueError):
+  """Point is incompatible with Elligator hashing"""
+
+def egcreate() -> Tuple[bytes, bytes]:
+  """
+  Create a random hidden key.
+  - Compatible with all of Ed25519, Curve25519 and Elligator2
+  - 254 bits of entropy (253 for Curve25519)
+
+  :returns: (hidden, edsk)
+  """
+  while True:
+    # Try until successful, half of our attempts should fail
+    with suppress(ElligatorError):
+      edsk = token_bytes(32)
+      return eghide(edsk), edsk
+
+def eghide(edsk) -> bytes:
+  """
+  Convert Ed25519 secret key into a random-looking 32-byte string.
+  - Deterministic, depends only on edsk
+
+  :raises ElligatorError: if the key is incompatible with Elligator2
+  """
+  # Calculate a dirty public key
+  s = dirty_scalar(edsk)
+  sg = s % 8  # sub group
+  # Using dirty generator: s * D - sg * G =
+  # Using normal generator: (s - sg) * G + LO[sg] =
+  # A dirty point produced: standard edpk + random low-order point
+  P = (s - sg) * G + LO[sg]
+  if not is_hashable(P.mont): raise ElligatorError("The key cannot be Elligator hashed")
+  # Take two pseudorandom bits (custom prefix needed to keep k and signatures secure)
+  # sha512(...)[31] & 0xC0 and placing at the same location on the final hidden byte.
+  tweak = sha(b"DirtyElligator2:" + edsk) & 0b11 << 254
+  # Elligator 2 hash
+  elligator = fast_curve_to_hash(P.mont, P.is_negative).val
```

Comment:
This right there introduces a bias that is conjectured to be too expensive to detect in practice. (Source: [comments from Mike Hamburg](https://old.reddit.com/r/crypto/comments/fd9t3m/elligator_with_x25519_is_broken_what_workaround/fjhemov/).)

It would be nice to add a comment to warn future reviewers about that. Otherwise it's all good.